### PR TITLE
Backport 2.28: Fix negative zero from bignum add/subtract

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
In the bignum module, operations of the form (-A) - (+A) or (-A) - (-A) with A > 0 created a "negative zero" representation of the value 0. Fix this.

The negative zero was processed correctly (i.e. as if it was a valid representation of 0) by most bignum operations (in particular all variants of add, sub, mul and div, as well as `mbedtls_mpi_mod_mpi`), but not all (`mbedtls_mpi_mod_int` outputs b instead of 0 for (-X) % b when X is a multiple of b, and `mbedtls_mpi_lt_mpi_ct` treats -0 as strictly less than +0). This bug had no consequence on cryptography code.

This is the 2.28 backport. I did it before the `development` version because in `development`, this may clash with the bignum refactoring.

Contents:

* Fix the bug.
* In bignum test case generation for binary operations, test both x op y and y op x. We previously refrained from doing that because it would have triggered this bug. This way we get a non-regression test.
* Start testing functions with a negative zero input. We don't guarantee that everything will work, but since we have a history of bugs, it's better to be robust. If there's an exception we don't want to handle (for example `mbedtls_mpi_lt_mpi_ct` would be annoying), we can add an exception when we get around to generating test cases for it.

## Gatekeeper checklist

- [x] **changelog** provided
- [x] **backport** (done before the main PR)
- [x] **tests** provided

### CODE REVIEW

File: tests/include/test/helpers.h [hunk 1]
```diff
@@ -360,20 +360,34 @@ void mbedtls_test_err_add_check( int high, int low,
...
- *
- * This is important so that the "0 (null)" and "0 (1 limb)" and
- * "leading zeros" test cases do what they claim.
+ * Like mbedtls_mpi_read_string(), but with tighter guarantees around
+ * edge cases.
+ *
+ * - This function guarantees that if \p s begins with '-' then the sign
+ *   bit of the result will be negative, even if the value is 0.
+ *   When this function encounters such a "negative 0", it
+ *   increments #mbedtls_test_read_mpi.
```
Comment (tom-cosgrove-arm) [inline tests/include/test/helpers.h:369]: should this be `mbedtls_test_case_uses_negative_0`?

Comment (gilles-peskine-arm) [reply tests/include/test/helpers.h:369]: Damn autocucumber!

This doesn't hinder review or CI so I'll push the fix later, combined with any other fixes.

File: library/bignum.c [hunk 2]
```diff
@@ -1249,27 +1249,34 @@ int mbedtls_mpi_sub_abs( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
...
+                        int flip_B )
 {
     int ret, s;
     MPI_VALIDATE_RET( X != NULL );
     MPI_VALIDATE_RET( A != NULL );
     MPI_VALIDATE_RET( B != NULL );
 
     s = A->s;
-    if( A->s * B->s < 0 )
+    if( A->s * B->s * flip_B < 0 )
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/bignum.c:1265]: Yes, although I'm always wary of bitwise operations with negative numbers since they can create trap representations, at least in theory. And I'm not convinced this is worth improving, especially in 2.28. And if it is, it's preexisting and feels like scope creep here.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/bignum.c:1265]: If we change it, I think we should change it everywhere there's a sign multiplication. The current code might be slightly suboptimal in performance, but it's at least consistent.

**Review by yanesca — DISMISSED** [review]
LGTM.

**Comment by tom-cosgrove-arm** [comment]
I have no other issues than the `mbedtls_test_read_mpi /mbedtls_test_case_uses_negative_0` typo - once that's fixed I will approve

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by yanesca — APPROVED** [review]
LGTM
